### PR TITLE
[pixart_rf] Add wireless devices

### PR DESCRIPTION
**Pull request by jyu99**
Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [x] Feature
- [x] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
There's a huge amount of duplication between the three new GObjects and we probably have to work out something better for that. Before we do that, can you describe why we need 3 different types? i.e. I understand the dongle and peripheral split, but why the difference between wireless-device and wireless-module?

**Comment by jyu99** [comment]
> There's a huge amount of duplication between the three new GObjects and we probably have to work out something better for that. Before we do that, can you describe why we need 3 different types? i.e. I understand the dongle and peripheral split, but why the difference between wireless-device and wireless-module?



My idea is wireless-device focus on response the get the modules information (actually it's dongle)

The wireless module is focus on OTA processing.

After get the information from wireless device , we can list modules on fwupd after fwupdtool get-devices 

there are different module. e.g  2.4G wireless mouse or keyboard

Here I define the dongle is kind of module. If we want to OTA dongle firmware ,use dongle module.

I concerned the 2.4G wireless device communication speed is much slower than USB dongle,so I use wireless-module to handle it.

I am not sure it's good or not about split wireless-device and dongle, perhaps they can combine into one file.

if you have any suggestion, please tell me. Thank you.

**Comment by hughsie** [comment]
I'm a bit confused; can you clarify a bit what you mean by modules and dongles? Any diagram on the logical relationship would be really useful. Thanks.

**Comment by jyu99** [comment]
image

**Comment by jyu99** [comment]
> I'm a bit confused; can you clarify a bit what you mean by modules and dongles? Any diagram on the logical relationship would be really useful. Thanks.

I have uploaded the diagram , but I am not sure that's what you want.

**Comment by hughsie** [comment]
> I have uploaded the diagram , but I am not sure that's what you want.

That's exactly what I wanted, thanks. I think in your diagram we would name things like this:

* Usb Dongle Receiver -> fu-pxi-receiver.c
* 2.4G wireless mouse -> fu-pxi-mouse-peripheral.c
* 2.4G wireless keyboard -> fu-pxi-keyboard-peripheral.c
* 2.4G wireless touch pad -> fu-pxi-touchpad-peripheral.c

Then, from an object point of view, FuPxiReceiver should derive from FuPxiDevice. FuPxiKeyboardPeripheral should derive from FuPxiPeripheral which then should also derive from FuPxiDevice. Then we have almost no code duplication, but still allowing different behaviours where required for kb/ms/kp.

**Comment by jyu99** [comment]
> > I have uploaded the diagram , but I am not sure that's what you want.
> 
> That's exactly what I wanted, thanks. I think in your diagram we would name things like this:
> 
> * Usb Dongle Receiver -> fu-pxi-receiver.c
> * 2.4G wireless mouse -> fu-pxi-mouse-peripheral.c
> * 2.4G wireless keyboard -> fu-pxi-keyboard-peripheral.c
> * 2.4G wireless touch pad -> fu-pxi-touchpad-peripheral.c
> 
> Then, from an object point of view, FuPxiReceiver should derive from FuPxiDevice. FuPxiKeyboardPeripheral should derive from FuPxiPeripheral which then should also derive from FuPxiDevice. Then we have almost no code duplication, but still allowing different behaviours where required for kb/ms/kp

> der



> > I have uploaded the diagram , but I am not sure that's what you want.
> 
> That's exactly what I wanted, thanks. I think in your diagram we would name things like this:
> 
> * Usb Dongle Receiver -> fu-pxi-receiver.c
> * 2.4G wireless mouse -> fu-pxi-mouse-peripheral.c
> * 2.4G wireless keyboard -> fu-pxi-keyboard-peripheral.c
> * 2.4G wireless touch pad -> fu-pxi-touchpad-peripheral.c
> 
> Then, from an object point of view, FuPxiReceiver should derive from FuPxiDevice. FuPxiKeyboardPeripheral should derive from FuPxiPeripheral which then should also derive from FuPxiDevice. Then we have almost no code duplication, but still allowing different behaviours where required for kb/ms/kp.

But FuPxiDevice is for BLE device using. I think  we should create a new object for handling dongle receiver.

That's why I create fu-pxi-wireless-devices.c   those wireless devices(mouse kb ,touch pad) OTA behavior are same,so I unify into fu-pxi-wireless-module.

**Comment by hughsie** [comment]
> I think we should create a new object for handling dongle receiver.

Okay, that seems okay.

> That's why I create fu-pxi-wireless-devices.c those wireless devices(mouse kb ,touch pad) OTA behavior are same,so I unify into fu-pxi-wireless-module.
> Maybe I can merge fu-pxi-wireless-devices.c and fu-pxi-dongle-module into fu-pxi-receiver.c, and rename the fu-pxi-wireless-module to fu-pxi-wireless-peripheral.

I think that's a good idea. Lets go with fu-pxi-peripheral.c and fu-pxi-receiver.c -- it's less typing.

File: plugins/pixart-rf/pixart-rf.quirk [hunk 1]
```diff
@@ -1,55 +1,70 @@
...
 # Chicony Cheetos mouse
 [HIDRAW\VEN_04F2&DEV_2103]
 Plugin = pixart_rf
+GType = FuPxiDevice
 
 # Primax Black Mocha KB
 [HIDRAW\VEN_0461&DEV_4EFA]
 Plugin = pixart_rf
+GType = FuPxiDevice
+
```
Comment (hughsie) [inline plugins/pixart-rf/pixart-rf.quirk:70]: Doesn't one of these need to be FuPxiWirelessPeripheral?

Comment (jyu99) [reply plugins/pixart-rf/pixart-rf.quirk:70]: No, It use FuPxiDevice. It's bluetooth device.

Comment (jyu99) [reply plugins/pixart-rf/pixart-rf.quirk:70]: Please tell me if you need any extra info. Thanks~

**Comment by superm1** [comment]
Can you please rebase this on master now?

**Comment by hughsie** [comment]
> Can you please rebase this on master now?

I think it's included in the recent merge. Please re-open a new PR if there are missing bits. Thanks!

**Pull request closed without merging**